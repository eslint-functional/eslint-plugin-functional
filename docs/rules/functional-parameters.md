# Enforce Functional Parameters (functional-parameters)

Disallow use of rest parameters, the `arguments` keyword and enforces that functions take a least 1 parameter.

## Rule Details

In functions, `arguments` is a special variable that is implicitly available.
This variable is an array containing the arguments passed to the function call and is often used to allow for any number of parameters to be passed to a function. Rest parameters are another way any number of parameters can be passed to a function.

When it comes to functional programming, known and explicit parameters must be used.

Note: With an unknown number of parameters, currying functions is a lot more difficult/impossible.

Examples of **incorrect** code for this rule:

```js
/* eslint functional/functional-parameters: "error" */

function add() {
  return arguments.reduce((sum, number) => sum + number, 0);
}
```

```js
/* eslint functional/functional-parameters: "error" */

function add(...numbers) {
  return numbers.reduce((sum, number) => sum + number, 0);
}
```

Examples of **correct** code for this rule:

```js
/* eslint functional/functional-parameters: "error" */

function add(numbers) {
  return numbers.reduce((sum, number) => sum + number, 0);
}
```

## Options

This rule accepts an options object of the following type:

```ts
{
  allowRestParameter: boolean;
  allowArgumentsKeyword: boolean;
  enforceParameterCount: false | "atLeastOne" | "exactlyOne" | {
    count: "atLeastOne" | "exactlyOne";
    ignoreIIFE: boolean;
  };
  ignorePattern?: string | Array<string>;
}
```

The default options:

```ts
{
  allowRestParameter: false,
  allowArgumentsKeyword: false,
  enforceParameterCount: {
    count: "atLeastOne",
    ignoreIIFE: true
  }
}
```

Note: the `lite` ruleset overrides the default options to:

```ts
{
  allowRestParameter: false,
  allowArgumentsKeyword: false,
  enforceParameterCount: false
}
```

### `allowRestParameter`

If true, this option allows for the use of rest parameters.

### `allowArgumentsKeyword`

If true, this option allows for the use of the `arguments` keyword.

### `enforceParameterCount`

This option enforces the number of parameters each function takes.

#### false

Don't enforce a parameter count.

#### "atLeastOne"

Require all functions to have a least one parameter.

There's not much point of having a function that doesn't take any parameters in functional programming.

#### "exactlyOne"

Require all functions to have exactly one parameter.

Any function that take takes multiple parameter can be rewritten as a higher-order function that only takes one.

Example:

```js
// This function
function add(x, y) {
  return x + y;
}

// can be rewritten like this.
function add(x) {
  return (y) => x + y;
}
```

See [Currying](https://en.wikipedia.org/wiki/Currying) and [Higher-order function](https://en.wikipedia.org/wiki/Higher-order_function) on Wikipedia for more infomation.

#### `enforceParameterCount.count`

See [enforceParameterCount](#enforceparametercount).

#### `enforceParameterCount.ignoreIIFE`

If true, this option allows for the use of [IIFEs](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) that do not have any parameters.

### `ignorePattern`

Patterns will be matched against function names.
See the [ignorePattern](./options/ignore-pattern.md) docs for more infomation.
