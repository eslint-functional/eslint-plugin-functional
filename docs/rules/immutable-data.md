# Disallow mutating objects and arrays (immutable-data)

This rule prohibits syntax that mutates existing objects and arrays via assignment to or deletion of their properties/elements.

## Rule Details

While requiring the `readonly` modifier forces declared types to be immutable,
it won't stop assignment into or modification of untyped objects or external types declared under different rules.

Examples of **incorrect** code for this rule:

```js
/* eslint functional/immutable-data: "error" */

const obj = { foo: 1 };

obj.foo += 2; // <- Modifying an existing object/array is not allowed.
obj.bar = 1; // <- Modifying an existing object/array is not allowed.
delete obj.foo; // <- Modifying an existing object/array is not allowed.
Object.assign(obj, { bar: 2 }); // <- Modifying properties of existing object not allowed.
```

```js
/* eslint functional/immutable-data: "error" */

const arr = [0, 1, 2];

arr[0] = 4; // <- Modifying an array is not allowed.
arr.length = 1; // <- Modifying an array is not allowed.
arr.push(3); // <- Modifying an array is not allowed.
```

Examples of **correct** code for this rule:

```js
/* eslint functional/immutable-data: "error" */

const obj = { foo: 1 };
const arr = [0, 1, 2];

const x = {
  ...obj
  bar: [
    ...arr, 3, 4
  ]
}
```

## Options

This rule accepts an options object of the following type:

```ts
{
  assumeTypes:
    | boolean
    | {
        forArrays: boolean;
        forObjects: boolean;
      }
  ignoreImmediateMutation: boolean;
  ignorePattern?: string | Array<string>;
  ignoreAccessorPattern?: string | Array<string>;
}
```

The default options:

```ts
{
  assumeTypes: true,
  ignoreImmediateMutation: true,
};
```

### `assumeTypes`

The rule take advantage of TypeScript's typing engine to check if mutation is taking place.
If you are not using TypeScript, type checking cannot be performed; hence this option exists.

This option will make the rule assume the type of the nodes it is checking are of type Array/Object.  
However this may result in some false positives being picked up.

Disabling this option can result in false negatives, for example:

```js
// When this option is DISABLED (and type info is not available).
const x = [0, 1, 2];
x.push(3); // This will NOT be flagged.
// This is due to the fact that without a typing engine, we cannot tell that x is an array.
```

Note: This option will have no effect if the TypeScript typing engine is avaliable (i.e. you are using TypeScript and have configured eslint correctly).

### `ignoreImmediateMutation`

If true, immediate mutation of objects before they are assigned to a variable is allowed.
This allows for the use of array mutator methods to be chained to newly created arrays.

For example, an array can be immutably sorted like so:

```js
const original = ["foo", "bar", "baz"];
const sorted = [...original].sort((a, b) => a.localeCompare(b)); // This is OK with ignoreImmediateMutation.
```

### `ignorePattern`

Patterns will be matched against variable names.
See the [ignorePattern](./options/ignore-pattern.md) docs for more infomation.

### `ignoreAccessorPattern`

This option takes a match string or an array of match strings (not a RegExp pattern).

The match string allows you to specify dot seperated `.` object paths and has support for "glob" `*` and "globstar" `**` matching.

For example:

```js
{
  // Ignore all reassigning to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*"
}
```

```js
{
  // Ignore all shallow mutations made to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*.*"
}
```

```js
{
  // Ignore all deep mutations made to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*.*.**"
}
```

```js
{
  // Ignore all deep mutations and reassigning to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*.**"
  // This is the same as `"ignoreAccessorPattern": ["**.mutable_*", "**.mutable_*.*.**"]`
}
```

#### Wildcards

The following wildcards can be used when specifing a pattern:

`**` - Match any depth (including zero). Can only be used as a full accessor.
`*` - When used as a full accessor, match the next accessor (there must be one). When used as part of an accessor, match any characters.
