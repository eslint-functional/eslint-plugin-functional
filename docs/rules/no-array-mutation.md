### no-array-mutation

[![Type Info Required][type-info-badge]][type-info-url]

This rule prohibits mutating an array via assignment to or deletion of their elements/properties. This rule enforces array immutability without the use of `ReadonlyArray<T>` (as apposed to [readonly-array](#readonly-array)).

```typescript
const x = [0, 1, 2];

x[0] = 4; // <- Mutating an array is not allowed.
x.length = 1; // <- Mutating an array is not allowed.
x.push(3); // <- Mutating an array is not allowed.
```

#### Has Fixer

No

#### Options

- [ignore-prefix](#using-the-ignore-prefix-option)
- [ignore-suffix](#using-the-ignore-suffix-option)
- [ignore-pattern](#using-the-ignore-pattern-option)
- [ignore-new-array](#using-the-ignore-new-array-option-with-no-array-mutation)
- ~~ignore-mutation-following-accessor~~ - _deprecated in favor of [ignore-new-array](#using-the-ignore-new-array-option-with-no-array-mutation)_

#### Example config

```javascript
"no-array-mutation": true
```

```javascript
"no-array-mutation": [true, {"ignore-prefix": "mutable"}]
```

```javascript
"no-array-mutation": [true, "ignore-new-array"]
```
