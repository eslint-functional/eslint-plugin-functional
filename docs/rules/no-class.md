# Disallow classes (no-class)

Disallow use of the `class` keyword.

## Rule Details

Thanks to libraries like [recompose](https://github.com/acdlite/recompose) and Redux's [React Container components](http://redux.js.org/docs/basics/UsageWithReact.html), there's not much reason to build Components using `React.createClass` or ES6 classes anymore. The `no-this` rule makes this explicit.

```typescript
const Message = React.createClass({
  render: function() {
    return <div>{this.props.message}</div>; // <- no this allowed
  }
});
```

Instead of creating classes, you should use React 0.14's [Stateless Functional Components](https://medium.com/@joshblack/stateless-components-in-react-0-14-f9798f8b992d#.t5z2fdit6) and save yourself some keystrokes:

```typescript
const Message = ({ message }) => <div>{message}</div>;
```

What about lifecycle methods like `shouldComponentUpdate`? We can use the [recompose](https://github.com/acdlite/recompose) library to apply these optimizations to your Stateless Functional Components. The [recompose](https://github.com/acdlite/recompose) library relies on the fact that your Redux state is immutable to efficiently implement shouldComponentUpdate for you.

```typescript
import { pure, onlyUpdateForKeys } from "recompose";

const Message = ({ message }) => <div>{message}</div>;

// Optimized version of same component, using shallow comparison of props
// Same effect as React's PureRenderMixin
const OptimizedMessage = pure(Message);

// Even more optimized: only updates if specific prop keys have changed
const HyperOptimizedMessage = onlyUpdateForKeys(["message"], Message);
```

## Options

The rule does not accept any options.
