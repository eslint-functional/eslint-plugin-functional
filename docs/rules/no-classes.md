<!-- markdownlint-disable -->
<!-- begin auto-generated rule header -->

# Disallow classes (`functional/no-classes`)

üíº This rule is enabled in the following configs: ‚òëÔ∏è `lite`, `noOtherParadigms`, ‚úÖ `recommended`, üîí `strict`.

<!-- end auto-generated rule header -->
<!-- markdownlint-restore -->
<!-- markdownlint-restore -->

Disallow use of the `class` keyword.

## Rule Details

### ‚ùå Incorrect

<!-- eslint-skip -->

```js
/* eslint functional/no-classes: "error" */

class Dog {
  constructor(name, age) {
    this.name = name;
    this.age = age;
  }

  get ageInDogYears() {
    return 7 * this.age;
  }
}

const dogA = new Dog("Jasper", 2);

console.log(`${dogA.name} is ${dogA.ageInDogYears} in dog years.`);
```

### ‚úÖ Correct

```js
/* eslint functional/no-classes: "error" */

function getAgeInDogYears(age) {
  return 7 * age;
}

const dogA = {
  name: "Jasper",
  age: 2,
};

console.log(`${dogA.name} is ${getAgeInDogYears(dogA.age)} in dog years.`);
```

### React Examples

Thanks to libraries like [recompose](https://github.com/acdlite/recompose) and Redux's
[React Container components](http://redux.js.org/docs/basics/UsageWithReact.html), there's not much reason to build
Components using `React.createClass` or ES6 classes anymore. The `no-this-expressions` rule makes this explicit.

```js
const Message = React.createClass({
  render() {
    return <div>{this.props.message}</div>; // <- no this allowed
  },
});
```

Instead of creating classes, you should use React 0.14's
[Stateless Functional
Components](https://medium.com/@joshblack/stateless-components-in-react-0-14-f9798f8b992d#.t5z2fdit6)
and save yourself some keystrokes:

```js
const Message = ({ message }) => <div>{message}</div>;
```

What about lifecycle methods like `shouldComponentUpdate`?
We can use the [recompose](https://github.com/acdlite/recompose) library to apply these optimizations to your
Stateless Functional Components. The [recompose](https://github.com/acdlite/recompose) library relies on the fact that
your Redux state is immutable to efficiently implement `shouldComponentUpdate` for you.

```js
import { onlyUpdateForKeys, pure } from "recompose";

const Message = ({ message }) => <div>{message}</div>;

// Optimized version of same component, using shallow comparison of props
// Same effect as React's PureRenderMixin
const OptimizedMessage = pure(Message);

// Even more optimized: only updates if specific prop keys have changed
const HyperOptimizedMessage = onlyUpdateForKeys(["message"], Message);
```
