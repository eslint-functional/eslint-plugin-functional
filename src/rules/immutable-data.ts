import type { TSESTree } from "@typescript-eslint/experimental-utils";
import { all as deepMerge } from "deepmerge";
import type { JSONSchema4 } from "json-schema";

import type {
  IgnoreAccessorPatternOption,
  IgnorePatternOption,
} from "~/common/ignore-options";
import {
  ignoreAccessorPatternOptionSchema,
  ignorePatternOptionSchema,
  shouldIgnore,
} from "~/common/ignore-options";
import { isExpected } from "~/utils/misc";
import type { RuleContext, RuleMetaData, RuleResult } from "~/utils/rule";
import { createRule, getTypeOfNode } from "~/utils/rule";
import { inConstructor } from "~/utils/tree";
import {
  isArrayConstructorType,
  isArrayExpression,
  isArrayType,
  isCallExpression,
  isIdentifier,
  isMemberExpression,
  isNewExpression,
  isObjectConstructorType,
} from "~/utils/typeguard";

// The name of this rule.
export const name = "immutable-data" as const;

// The options this rule can take.
type Options = IgnorePatternOption &
  IgnoreAccessorPatternOption & {
    readonly ignoreImmediateMutation: boolean;
    readonly assumeTypes:
      | boolean
      | {
          readonly forArrays: boolean;
          readonly forObjects: boolean;
        };
  };

// The schema for the rule options.
const schema: JSONSchema4 = [
  deepMerge([
    ignorePatternOptionSchema,
    ignoreAccessorPatternOptionSchema,
    {
      type: "object",
      properties: {
        ignoreImmediateMutation: {
          type: "boolean",
        },
        assumeTypes: {
          oneOf: [
            {
              type: "boolean",
            },
            {
              type: "object",
              properties: {
                forArrays: {
                  type: "boolean",
                },
                forObjects: {
                  type: "boolean",
                },
              },
              additionalProperties: false,
            },
          ],
        },
      },
      additionalProperties: false,
    },
  ]),
];

// The default options for the rule.
const defaultOptions: Options = {
  ignoreImmediateMutation: true,
  assumeTypes: {
    forArrays: true,
    forObjects: true,
  },
};

// The possible error messages.
const errorMessages = {
  generic: "Modifying an existing object/array is not allowed.",
  object: "Modifying properties of existing object not allowed.",
  array: "Modifying an array is not allowed.",
} as const;

// The meta data for this rule.
const meta: RuleMetaData<keyof typeof errorMessages> = {
  type: "suggestion",
  docs: {
    description: "Enforce treating data as immutable.",
    category: "Best Practices",
    recommended: "error",
  },
  messages: errorMessages,
  schema,
};

/**
 * Array methods that mutate an array.
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/prototype#Methods#Mutator_methods
 */
const arrayMutatorMethods = [
  "copyWithin",
  "fill",
  "pop",
  "push",
  "reverse",
  "shift",
  "sort",
  "splice",
  "unshift",
] as const;

/**
 * Array methods that return a new object (or array) without mutating the original.
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/prototype#Methods#Accessor_methods
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/prototype#Iteration_methods
 */
const arrayNewObjectReturningMethods = [
  "concat",
  "slice",
  "filter",
  "map",
  "reduce",
  "reduceRight",
] as const;

/**
 * Array constructor functions that create a new array.
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array#Methods
 */
const arrayConstructorFunctions = ["from", "of"] as const;

/**
 * Object constructor functions that mutate an object.
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object#Methods_of_the_Object_constructor
 */
const objectConstructorMutatorFunctions = [
  "assign",
  "defineProperties",
  "defineProperty",
  "setPrototypeOf",
] as const;

/**
 * Check if the given assignment expression violates this rule.
 */
function checkAssignmentExpression(
  node: TSESTree.AssignmentExpression,
  context: RuleContext<keyof typeof errorMessages, Options>
): RuleResult<keyof typeof errorMessages, Options> {
  return {
    context,
    descriptors:
      isMemberExpression(node.left) &&
      // Allow if in a constructor - allow for field initialization.
      !inConstructor(node)
        ? [{ node, messageId: "generic" }]
        : [],
  };
}

/**
 * Check if the given node violates this rule.
 */
function checkUnaryExpression(
  node: TSESTree.UnaryExpression,
  context: RuleContext<keyof typeof errorMessages, Options>
): RuleResult<keyof typeof errorMessages, Options> {
  return {
    context,
    descriptors:
      node.operator === "delete" && isMemberExpression(node.argument)
        ? [{ node, messageId: "generic" }]
        : [],
  };
}

/**
 * Check if the given node violates this rule.
 */
function checkUpdateExpression(
  node: TSESTree.UpdateExpression,
  context: RuleContext<keyof typeof errorMessages, Options>,
  options: Options
): RuleResult<keyof typeof errorMessages, Options> {
  return {
    context,
    descriptors:
      isMemberExpression(node.argument) &&
      !shouldIgnore(node.argument, context, options)
        ? [{ node, messageId: "generic" }]
        : [],
  };
}

/**
 * Check if the given the given MemberExpression is part of a chain and
 * immediately follows a method/function call that returns a new array.
 *
 * If this is the case, then the given MemberExpression is allowed to be
 * a mutator method call.
 */
function isInChainCallAndFollowsNew(
  node: TSESTree.MemberExpression,
  context: RuleContext<keyof typeof errorMessages, Options>,
  assumeArrayTypes: boolean
): boolean {
  return (
    // Check for: [0, 1, 2]
    isArrayExpression(node.object) ||
    // Check for: new Array()
    (isNewExpression(node.object) &&
      isArrayConstructorType(
        // `isNewExpression` type guard doesn't seem to be working? so use `as`.
        getTypeOfNode((node.object as TSESTree.NewExpression).callee, context),
        assumeArrayTypes,
        (node.object as TSESTree.NewExpression).callee
      )) ||
    (isCallExpression(node.object) &&
      isMemberExpression(node.object.callee) &&
      isIdentifier(node.object.callee.property) &&
      // Check for: Array.from(iterable)
      ((arrayConstructorFunctions.some(
        isExpected(node.object.callee.property.name)
      ) &&
        isArrayConstructorType(
          getTypeOfNode(node.object.callee.object, context),
          assumeArrayTypes,
          node.object.callee.object
        )) ||
        // Check for: array.slice(0)
        arrayNewObjectReturningMethods.some(
          isExpected(node.object.callee.property.name)
        )))
  );
}

/**
 * Check if the given node violates this rule.
 */
function checkCallExpression(
  node: TSESTree.CallExpression,
  context: RuleContext<keyof typeof errorMessages, Options>,
  options: Options
): RuleResult<keyof typeof errorMessages, Options> {
  const assumeTypesForArrays =
    options.assumeTypes === true ||
    (options.assumeTypes !== false && options.assumeTypes.forArrays === true);
  const assumeTypesForObjects =
    options.assumeTypes === true ||
    (options.assumeTypes !== false && options.assumeTypes.forObjects === true);

  const isMethodName = (methodName) => {
    return (
      methodName ===
      ((node.callee as TSESTree.MemberExpression)
        .property as TSESTree.Identifier).name
    );
  };

  return {
    context,
    descriptors:
      // Potential object mutation?
      isMemberExpression(node.callee) && isIdentifier(node.callee.property)
        ? // Potential array mutation?
          // Check if allowed here - this cannot be automatically checked beforehand.
          !shouldIgnore(node.callee.object, context, options) &&
          arrayMutatorMethods.some(isMethodName) &&
          (!options.ignoreImmediateMutation ||
            !isInChainCallAndFollowsNew(
              node.callee,
              context,
              assumeTypesForArrays
            )) &&
          isArrayType(
            getTypeOfNode(node.callee.object, context),
            assumeTypesForArrays,
            node.callee.object
          )
          ? [{ node, messageId: "array" }]
          : // Potential non-array object mutation (ex. Object.assign on identifier)?
          objectConstructorMutatorFunctions.some(isMethodName) &&
            node.arguments.length >= 2 &&
            (isIdentifier(node.arguments[0]) ||
              isMemberExpression(node.arguments[0])) &&
            // Check if allowed here - this cannot be automatically checked beforehand.
            !shouldIgnore(node.arguments[0], context, options) &&
            isObjectConstructorType(
              getTypeOfNode(node.callee.object, context),
              assumeTypesForObjects,
              node.callee.object
            )
          ? [{ node, messageId: "object" }]
          : []
        : [],
  };
}

// Create the rule.
export const rule = createRule<keyof typeof errorMessages, Options>(
  name,
  meta,
  defaultOptions,
  {
    AssignmentExpression: checkAssignmentExpression,
    UnaryExpression: checkUnaryExpression,
    UpdateExpression: checkUpdateExpression,
    CallExpression: checkCallExpression,
  }
);
