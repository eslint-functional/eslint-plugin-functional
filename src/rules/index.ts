import * as functionalParameters from "./functional-parameters";
import * as immutableData from "./immutable-data";
import * as noClass from "./no-class";
import * as noConditionalStatement from "./no-conditional-statement";
import * as noExpressionStatement from "./no-expression-statement";
import * as noLet from "./no-let";
import * as noLoop from "./no-loop-statement";
import * as noMethodSignature from "./no-method-signature";
import * as noMixedType from "./no-mixed-type";
import * as noPromiseReject from "./no-promise-reject";
import * as noReturnVoid from "./no-return-void";
import * as noThisExpression from "./no-this-expression";
import * as noThrowStatement from "./no-throw-statement";
import * as noTryStatement from "./no-try-statement";
import * as preferReadonlyTypes from "./prefer-readonly-type";
import * as preferType from "./prefer-type-literal";

/**
 * All of the custom rules.
 */
export const rules = {
  [functionalParameters.name]: functionalParameters.rule,
  [immutableData.name]: immutableData.rule,
  [noClass.name]: noClass.rule,
  [noConditionalStatement.name]: noConditionalStatement.rule,
  [noExpressionStatement.name]: noExpressionStatement.rule,
  [noLet.name]: noLet.rule,
  [noLoop.name]: noLoop.rule,
  [noMethodSignature.name]: noMethodSignature.rule,
  [noMixedType.name]: noMixedType.rule,
  [noPromiseReject.name]: noPromiseReject.rule,
  [noReturnVoid.name]: noReturnVoid.rule,
  [noThisExpression.name]: noThisExpression.rule,
  [noThrowStatement.name]: noThrowStatement.rule,
  [noTryStatement.name]: noTryStatement.rule,
  [preferReadonlyTypes.name]: preferReadonlyTypes.rule,
  [preferType.name]: preferType.rule,
};
