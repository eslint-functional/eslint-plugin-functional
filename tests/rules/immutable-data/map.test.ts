import dedent from "dedent";
import { createRuleTester } from "eslint-vitest-rule-tester";
import { describe, expect, it } from "vitest";

import { name, rule } from "#/rules/immutable-data";

import { typescriptConfig } from "../../utils/configs";

describe(name, () => {
  describe("typescript", () => {
    const { valid, invalid } = createRuleTester({
      name,
      rule,
      configs: typescriptConfig,
    });

    it("report mutating map methods", async () => {
      const invalidResult = await invalid({
        code: dedent`
          const x = new Map([[5, 6]]);
          x.set(4, 8);
          x.delete(4);
          x.clear();
        `,
        options: [],
        errors: ["map", "map", "map"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it("doesn't report mutating map methods when ignoring maps and sets", async () => {
      await valid({
        code: dedent`
          const x = new Map([[5, 6]]);
          x.set(4, 8);
          x.delete(4);
          x.clear();
        `,
        options: [
          {
            ignoreMapsAndSets: true,
          },
        ],
      });
    });

    it("doesn't report non-mutating map methods", async () => {
      await valid(dedent`
        const x = new Map([[5, 6]]);
        x.size;
        x.has(4);
        x.values();
        x.entries();
        x.keys();
      `);
    });

    it("doesn't report mutating map methods on non-map objects", async () => {
      await valid(dedent`
        const z = {
          set: function () {},
          delete: function () {},
          clear: function () {},
        };

        z.set();
        z.delete();
        z.clear();
      `);
    });

    describe("options", () => {
      describe("ignoreNonConstDeclarations", () => {
        it("reports variables declared as const", async () => {
          const invalidResult = await invalid({
            code: dedent`
              const x = new Map([[5, 6]]);
              x.set(4, 8);
              x.delete(4);
              x.clear();
            `,
            options: [
              {
                ignoreNonConstDeclarations: true,
              },
            ],
            errors: ["map", "map", "map"],
          });
          expect(invalidResult.result).toMatchSnapshot();
        });

        it("doesn't report variables not declared as const", async () => {
          await valid({
            code: dedent`
              let x = new Map([[5, 6]]);
              x.set(4, 8);
              x.delete(4);
              x.clear();
            `,
            options: [
              {
                ignoreNonConstDeclarations: true,
              },
            ],
          });
        });
      });

      describe("ignoreImmediateMutation", () => {
        it("doesn't report immediately mutation when enabled", async () => {
          await valid({
            code: dedent`
              new Map([[5, 6]]).set(4, 8);
              new Map([[5, 6]]).delete(4);
              new Map([[5, 6]]).clear();
            `,
            options: [
              {
                ignoreImmediateMutation: true,
              },
            ],
          });
        });

        it("reports immediately mutation when disabled", async () => {
          const invalidResult = await invalid({
            code: dedent`
              new Map([[5, 6]]).set(4, 8);
              new Map([[5, 6]]).delete(4);
              new Map([[5, 6]]).clear();
            `,
            options: [
              {
                ignoreImmediateMutation: false,
              },
            ],
            errors: ["map", "map", "map"],
          });
          expect(invalidResult.result).toMatchSnapshot();
        });
      });
    });
  });
});
