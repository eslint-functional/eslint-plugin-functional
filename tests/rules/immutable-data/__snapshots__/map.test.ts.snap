// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`immutable-data > typescript > options > ignoreImmediateMutation > reports immediately mutation when disabled 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 1,
      "endColumn": 28,
      "endLine": 1,
      "line": 1,
      "message": "Modifying a map is not allowed.",
      "messageId": "map",
      "nodeType": "CallExpression",
      "ruleId": "immutable-data",
      "severity": 2,
    },
    {
      "column": 1,
      "endColumn": 28,
      "endLine": 2,
      "line": 2,
      "message": "Modifying a map is not allowed.",
      "messageId": "map",
      "nodeType": "CallExpression",
      "ruleId": "immutable-data",
      "severity": 2,
    },
    {
      "column": 1,
      "endColumn": 26,
      "endLine": 3,
      "line": 3,
      "message": "Modifying a map is not allowed.",
      "messageId": "map",
      "nodeType": "CallExpression",
      "ruleId": "immutable-data",
      "severity": 2,
    },
  ],
  "output": "new Map([[5, 6]]).set(4, 8);
new Map([[5, 6]]).delete(4);
new Map([[5, 6]]).clear();",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 1,
          "endColumn": 28,
          "endLine": 1,
          "line": 1,
          "message": "Modifying a map is not allowed.",
          "messageId": "map",
          "nodeType": "CallExpression",
          "ruleId": "rule-to-test/immutable-data",
          "severity": 2,
        },
        {
          "column": 1,
          "endColumn": 28,
          "endLine": 2,
          "line": 2,
          "message": "Modifying a map is not allowed.",
          "messageId": "map",
          "nodeType": "CallExpression",
          "ruleId": "rule-to-test/immutable-data",
          "severity": 2,
        },
        {
          "column": 1,
          "endColumn": 26,
          "endLine": 3,
          "line": 3,
          "message": "Modifying a map is not allowed.",
          "messageId": "map",
          "nodeType": "CallExpression",
          "ruleId": "rule-to-test/immutable-data",
          "severity": 2,
        },
      ],
      "output": "new Map([[5, 6]]).set(4, 8);
new Map([[5, 6]]).delete(4);
new Map([[5, 6]]).clear();",
    },
  ],
}
`;

exports[`immutable-data > typescript > options > ignoreNonConstDeclarations > reports variables declared as const 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 1,
      "endColumn": 12,
      "endLine": 2,
      "line": 2,
      "message": "Modifying a map is not allowed.",
      "messageId": "map",
      "nodeType": "CallExpression",
      "ruleId": "immutable-data",
      "severity": 2,
    },
    {
      "column": 1,
      "endColumn": 12,
      "endLine": 3,
      "line": 3,
      "message": "Modifying a map is not allowed.",
      "messageId": "map",
      "nodeType": "CallExpression",
      "ruleId": "immutable-data",
      "severity": 2,
    },
    {
      "column": 1,
      "endColumn": 10,
      "endLine": 4,
      "line": 4,
      "message": "Modifying a map is not allowed.",
      "messageId": "map",
      "nodeType": "CallExpression",
      "ruleId": "immutable-data",
      "severity": 2,
    },
  ],
  "output": "const x = new Map([[5, 6]]);
x.set(4, 8);
x.delete(4);
x.clear();",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 1,
          "endColumn": 12,
          "endLine": 2,
          "line": 2,
          "message": "Modifying a map is not allowed.",
          "messageId": "map",
          "nodeType": "CallExpression",
          "ruleId": "rule-to-test/immutable-data",
          "severity": 2,
        },
        {
          "column": 1,
          "endColumn": 12,
          "endLine": 3,
          "line": 3,
          "message": "Modifying a map is not allowed.",
          "messageId": "map",
          "nodeType": "CallExpression",
          "ruleId": "rule-to-test/immutable-data",
          "severity": 2,
        },
        {
          "column": 1,
          "endColumn": 10,
          "endLine": 4,
          "line": 4,
          "message": "Modifying a map is not allowed.",
          "messageId": "map",
          "nodeType": "CallExpression",
          "ruleId": "rule-to-test/immutable-data",
          "severity": 2,
        },
      ],
      "output": "const x = new Map([[5, 6]]);
x.set(4, 8);
x.delete(4);
x.clear();",
    },
  ],
}
`;

exports[`immutable-data > typescript > report mutating map methods 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 1,
      "endColumn": 12,
      "endLine": 2,
      "line": 2,
      "message": "Modifying a map is not allowed.",
      "messageId": "map",
      "nodeType": "CallExpression",
      "ruleId": "immutable-data",
      "severity": 2,
    },
    {
      "column": 1,
      "endColumn": 12,
      "endLine": 3,
      "line": 3,
      "message": "Modifying a map is not allowed.",
      "messageId": "map",
      "nodeType": "CallExpression",
      "ruleId": "immutable-data",
      "severity": 2,
    },
    {
      "column": 1,
      "endColumn": 10,
      "endLine": 4,
      "line": 4,
      "message": "Modifying a map is not allowed.",
      "messageId": "map",
      "nodeType": "CallExpression",
      "ruleId": "immutable-data",
      "severity": 2,
    },
  ],
  "output": "const x = new Map([[5, 6]]);
x.set(4, 8);
x.delete(4);
x.clear();",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 1,
          "endColumn": 12,
          "endLine": 2,
          "line": 2,
          "message": "Modifying a map is not allowed.",
          "messageId": "map",
          "nodeType": "CallExpression",
          "ruleId": "rule-to-test/immutable-data",
          "severity": 2,
        },
        {
          "column": 1,
          "endColumn": 12,
          "endLine": 3,
          "line": 3,
          "message": "Modifying a map is not allowed.",
          "messageId": "map",
          "nodeType": "CallExpression",
          "ruleId": "rule-to-test/immutable-data",
          "severity": 2,
        },
        {
          "column": 1,
          "endColumn": 10,
          "endLine": 4,
          "line": 4,
          "message": "Modifying a map is not allowed.",
          "messageId": "map",
          "nodeType": "CallExpression",
          "ruleId": "rule-to-test/immutable-data",
          "severity": 2,
        },
      ],
      "output": "const x = new Map([[5, 6]]);
x.set(4, 8);
x.delete(4);
x.clear();",
    },
  ],
}
`;
